common/src/primitives/merkle/tree.rs

```diff
@@ -0,0 +1,560 @@
+        //       on creation by taking Vec<H256> as input and padding it and wrapping it with a strong type
+        if leaves.is_empty() {
+            return Err(MerkleTreeFormError::TooSmall(leaves.len()));
+        }
+        let padding = create_merkletree_padding(&leaves);
+        let leaves = leaves.into_iter().chain(padding).collect::<Vec<_>>();
+        let steps = leaves.len() - 1;
+        let mut tree = Vec::with_capacity(2 * leaves.len() - 1);
+        tree.extend(leaves.into_iter());
+        for i in 0..steps {
+            let el = Self::combine_pair(&tree[i * 2], &tree[i * 2 + 1]);
+            tree.push(el);
+        }
+        let res = Self { tree };
+        Ok(res)
+    }
+
+    /// Get the root of the merkle tree.
+    pub fn root(&self) -> H256 {
+        *self.tree.last().expect("By design, at least one element must exist")
+    }
+
+    fn leaves_count_from_tree_size(
+        tree_size: NonZeroUsize,
+    ) -> Result<NonZeroUsize, MerkleTreeProofExtractionError> {
+        if (tree_size.get() + 1).count_ones() != 1 {
+            return Err(MerkleTreeProofExtractionError::InvalidTreeSize(
+                tree_size.get(),
+            ));
+        }
+
+        let tree_size = tree_size.get();
+        let leaves_count = (tree_size + 1) >> 1;
+        debug_assert!(leaves_count.is_power_of_two());
+        Ok(NonZeroUsize::new(leaves_count).expect("By design, tree_size is always > 0"))
+    }
+
+    /// Ensure the leaves indices are sorted and unique
+    fn is_sorted_and_unique(leaves_indices: &[u32]) -> bool {
+        let vals: BTreeSet<u32> = leaves_indices.iter().copied().collect();
+        let vals: Vec<u32> = vals.into_iter().collect();
+        vals == leaves_indices
+    }
+
+    /// Find adjacent leaves indices in a merkle tree
+    fn get_adjacent_indices_states(
+        leaves_indices: &[u32],
+    ) -> Result<Vec<AdjacentLeavesIndices>, MerkleTreeProofExtractionError> {
+        if !Self::is_sorted_and_unique(leaves_indices) {
+            return Err(
+                MerkleTreeProofExtractionError::UnsortedOrUniqueLeavesIndices(
+                    leaves_indices.to_vec(),
+                ),
+            );
+        }
+
+        let mut res = Vec::with_capacity(leaves_indices.len());
+
+        // we chain the windows with a max value to ensure we get the last element if it doesn't pair with the preceding value
+        let max_chain = std::iter::once(&u32::MAX);
+        for win in leaves_indices.iter().chain(max_chain).tuple_windows::<(&u32, &u32)>() {
+            let (a, b) = (*win.0, *win.1);
+
+            // In a tree, we expect elements to be adjacent if the first one has even index and the second one has index + 1.
+            if a % 2 == 0 && a + 1 == b {
+                res.push(AdjacentLeavesIndices::Together(a as usize, b as usize));
+            } else {
+                res.push(AdjacentLeavesIndices::Alone(a as usize));
+            }
+        }
```

Comment:
Looking at this, for example the indices [0,1,2] will become windows [(0, 1), (1, 2), (2, MAX)]. So the result will have [Together(0, 1), Alone(1), Alone(2)]. Index 1 is included twice. Is that intentional? Or did I miss / misunderstand something?